src/terminal.rs

```diff
@@ -0,0 +1,78 @@
+use std::io::{Stdin, Stdout, Write};
+use crate::terminalerror::TerminalError;
+use crate::todo::Todo;
+use console::style;
+
+pub struct Terminal {
+    stdin: Stdin,
+    stdout: Stdout,
+}
+
+impl Terminal {
+    pub fn new() -> Self {
+        Terminal {
+            stdin: std::io::stdin(),
+            stdout: std::io::stdout(),
+        }
+    }
+
+    fn input(&mut self) -> Result<String, TerminalError> {
+        let mut buf: String = String::new();
+
+        match self.stdin.read_line(&mut buf) {
+            Ok(_) => Ok(buf.trim().to_string()),
+            Err(error) => Err(TerminalError::Stdin(error))
+        }
+    }
+    fn receive_option(&mut self) -> Result<String, TerminalError> {
+        let mut res = self.input()?;
+        if res.to_lowercase() == "sim" {
+            return Ok(res)
+        } else if res.to_lowercase() == "nao" {
+            return Ok(res)
+        } else {
+            loop {
+                println!("{}", style("COMANDO ERRADO").red());
+                println!("Digite {} ou {}", style("Sim").blue().bold(), style("Nao").yellow().bold());
+                res = self.input()?;
+                if res.to_lowercase() == "sim" || res.to_lowercase() == "nao" {
+                    break;
+                }
+            }
+        }
+        Ok(res)
```

Comment:
Podemos melhorar bastante esse parte!

Note que os retornos sempre são os mesmos, então podemos ajustar a lógica dos condicionais para mantermos apenas o `loop`, que podemos simplificar ainda mais juntando o condicional inicial junto com o condicional do `break` em um `while`.

`matches!(<expr>, <pattern>)` é uma macro que verifica se o valor de `<expr>` satisfaz o pattern `<pattern>`, retornando um booleano. É equivalente a:

```rust
match <expr> {
    <pattern> => true,
    _ => false,
}
```

Perceba que com essas mudanças a lógica da função é muito mais clara.

```suggestion
        while !matches!(res.to_lowercase(), "sim" | "nao") {
            println!("{}", style("COMANDO ERRADO").red());
            println!("Digite {} ou {}", style("Sim").blue().bold(), style("Nao").yellow().bold());
            res = self.input()?;
        }
        
        Ok(res)
```

Reply:
Tá, vou estudar pra tentar entender bem! Obrigado pela sugestão!

Reply:
Eu copiei a sua sugestão ai mas tava com erro no código, mas daí refiz com base na sua referência e consegui usar uma lógica do while, acredito que ficou melhor!

Reply:
Para usar o match aqui, vc precisa converter a `String` para um `&str`. Vc pode fazer isso com `&*res`, por razões que vão ser explicadas no futuro :x

Reply:
Ou usando `res.as_str()`, mas `&*res` é mais comum.